include/git2/status.h

```diff
@@ -294,6 +299,31 @@ GIT_EXTERN(int) git_status_file(
 	const char *path);
 
 /**
+ * Get file status for a single file, relative to a specified tree.
+ *
+ * This works like `git_status_file`, except it allows you to specify a tree
+ * other than the head commit's.
+ *
+ * Calling this with `head_tree` set to NULL is the same as calling
+ * `git_status_file`.
+ *
+ * @param status_flags Output combination of git_status_t values for file
+ * @param repo A repository object
+ * @param path The exact path to retrieve status for relative to the
+ * repository working directory
+ * @param head_tree The tree to be used instead of the head commit's tree,
+ * for comparing to the indxex.
+ * @return 0 on success, GIT_ENOTFOUND if the file is not found in the
+ *      specified tree, index, and work tree, GIT_EAMBIGUOUS if `path` matches
+ *      multiple files or if it refers to a folder, and -1 on other errors.
+ */
+GIT_EXTERN(int) git_status_file_at(
+	unsigned int *status_flags,
+	git_repository *repo,
+	const char *path,
+	git_tree *head_tree);
```

Comment:
So one problem with this new API is, as @ethomson already mentioned, that we still have to add new calls whenever we want some new option in here. Ideally, we'd have designed `git_status_file` to just take a `git_status_options` structure here, so that we can add more options later on without adding new functions for all of them.

As we do not want to break the API here, we obviously cannot change `git_status_file` right now. What we _can_ do though is to add a `git_status_file_ext(unsigned int *, git_repository *, const char *, const git_status_options *)`, which fixes that shortcoming. See for example `git_status_foreach` and `git_status_foreach_ext`, where we fixed the same shortcoming.

So I'd prefer to introduce that one instead of having to potentially maintain more `git_status_file_foobar` functions in the future.

Reply:
Actually, I was suggesting not adding a new `git_status_file` API _at all_.  I was hoping that the change in the options struct would be enough.

`git_status_file` is a helper method.  It should be for the 99% use cases.  If you want something special (like a different `HEAD` tree, which is _super_ uncommon) then I don't mind saying that you have to do a little more work to get there.

In other words, I am not in favor of adding new `git_status_file`-like functions, and would encourage you to just call `git_status_foreach_ext` (with the new opts structure) for this very special use case.

Reply:
So you think I should take the login from `git_status_file_at` and move it up into my app? Either way is fine with me.

Reply:
Well, if you have `git_status_foreach_ext` handle the new tree option then most of it is already done for you, I guess. But otherwise yes